Update serialization mechanism

gmazzap opened the pull request:
Currently we have [a rule that discourages the usage of magic `__serialize` and `__unserialize`](https://github.com/inpsyde/php-coding-standards/blob/version/2/Inpsyde/Sniffs/CodeQuality/DisableMagicSerializeSniff.php#L16-L18) while suggesting usage of the `Serializable` interface.

That was the "state of the art" up to PHP 7.3, and the rule even predates that.

With PHP 7.4 the `Serializable` interface is [not recommended anymore](https://wiki.php.net/rfc/custom_object_serialization) (the RFC define it as _"severely broken"_).

The new recommended method is to use `__serialize` and `__unserialize` methods.

Not being a major release PHP 7.4 did not deprecate `Serializable`.

PHP 8.1 [triggers a deprecation warning](ttps://www.php.net/manual/en/class.serializable.php) if `Serializable` without _also_ having `__serialize` and `__unserialize`.

And PHP 9 will probably remove the interface altogether.

Because v2 is PHP 7.4 we can safely discourage `Serializable` already and fully recommend `__serialize` and `__unserialize`.

This PR does that by:

- Edit the sniff that discouraged `__serialize` and `__unserialize` to actually suggest them where `__sleep` and `__wakeup` are used
- Add a sniff to discourage `Serializable` interface, suggesting `__serialize` and `__unserialize` 

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)


**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Feature.


**What is the current behavior?** (You can also link to an open issue here)

- `__serialize` and `__unserialize` are discouraged
- `Serializable` interface is suggested

**What is the new behavior (if this is a feature change)?**

- `__serialize` and `__unserialize` are suggested
- `Serializable` interface is discouraged

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

- New errors might be reported

**Other information**:

--

gmazzap committed:
Update serialization mechanism

- Suggest suggest __serialize and __unserialize (previously discouraged)
- Add sniff to discourage Serializable interface

See
- https://wiki.php.net/rfc/custom_object_serialization
- https://www.php.net/manual/en/class.serializable.php

gmazzap committed:
Fix QA

tfrommen commented on Inpsyde/Sniffs/CodeQuality/DisableMagicSerializeSniff.php:
```diff
@@ -45,7 +43,8 @@ public function process(File $phpcsFile, $stackPtr): void
                 sprintf(
-                    'The method "%s" is forbidden, please use Serializable interface.',
+                    'The method "%s" is deprecated, '
+                    . 'please use __serialize and __unserialize instead.',
```
This will render one of the following:

- The method "__sleep" is deprecated, please use __serialize and __unserialize instead.
- The method "__wakeup" is deprecated, please use __serialize and __unserialize instead.

Wouldnt it be better to use a map so that people see the exact "replacement"?

```php
[
    '__sleep' => '__serialize',
    '__wakeup' => '__unserialize',
]
```

gmazzap replied:
Thank you, [done](https://github.com/inpsyde/php-coding-standards/pull/90/commits/7104de07a3f0773966c543cb6019fd3c17b55a82).

shvlv reviewed (approved):
We have to update documentation https://github.com/inpsyde/php-coding-standards/blob/85e25cf602f75c000434a5693a024abc2d09bc5d/README.md#L124 and add a new item for `DisableSerializeInterfaceSniff`.

Apart from that looks good :+1:

gmazzap committed:
Update documentation for serialization-related sniffs

- Added docs for DisableSerializeInterface
- Updated docs for DisableMagicSerialize

gmazzap committed:
Improve DisableMagicSerializeSniff error message

gmazzap commented:
> We have to update documentation

Thank you @shvlv, [done](https://github.com/inpsyde/php-coding-standards/pull/90/commits/e77a9e31e5afd515444b8a96d8a0357c58776e3e).
